Implement Slashing Protection

pscott opened the pull request:
## Issue Addressed

Closes #254 
Closes #623 

## Proposed Changes

- Add a slashing_protection crate in validator_client.

## Additional info
Added some comments on github for things I'm unsure about.

Please see this [hackmd note](https://hackmd.io/@6Uku5jlsSVewmiY5aPVYIA/HJO_JZmKB) that explains the slashing protection implementation.

pscott commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+        check_for_attester_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
```
I would've liked to have this method be a generic one, but I didn't manage to do so

pscott commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+        check_for_proposer_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
```
Would've like to have this method be generic but didn't manage to do so

pscott commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+        let mut file = match File::open(filepath) {
+            Ok(file) => file,
+            Err(e) => match e.kind() {
+                // File was not found meaning it has not been created yet, and that history is empty.
```
Are we sure this is the behaviour we want? Or would we rather create the file in advance, and returning an error here if file was not found (i.e assuming that the file already exists)

pscott commented on validator_client/src/attestation_producer/mod.rs:
```diff
@@ -143,21 +142,8 @@ impl<'a, B: BeaconNodeAttestation, S: Signer, E: EthSpec> AttestationProducer<'a
-    /// Important: this function is presently stubbed-out. It provides ZERO SAFETY.
-    fn store_produce(&mut self, _attestation: &AttestationData) {
-        // TODO: Implement slash protection
+    fn safe_to_produce(&self, attestation: &AttestationData) -> bool {
```
Having it return a bool here means we can't have it return the extra info (error info or validity reason). As the code was already written I didn't change it, but I would argue returning an enum and using it upstream would be better.

pscott commented on validator_client/src/block_producer/mod.rs:
```diff
@@ -141,24 +139,9 @@ impl<'a, B: BeaconNodeBlock, S: Signer, E: EthSpec> BlockProducer<'a, B, S, E> {
-    fn store_produce(&mut self, _block: &BeaconBlock<E>) {
-        // TODO: record this block production to prevent future slashings.
-        // https://github.com/sigp/lighthouse/issues/160
+    fn safe_to_produce(&self, block: &BeaconBlock<E>) -> bool {
```
Having it return a bool here means we can't have it return the extra info (error info or validity reason). As the code was already written I didn't change it, but I would argue returning an enum and using it upstream would be better.

pscott commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+impl<T: Encode + Decode + Clone + PartialEq> HistoryInfo<T> {
+    /// Inserts the incoming data in the in-memory history, and writes it to the history file.
+    pub fn insert_and_write(&mut self, data: T, index: usize) -> IOResult<()> {
+        self.data.insert(index, data);
```
should we `assert!(index < data.len())` ?

paulhauner replied:
Since this is a public function, I think we should return an error if it will cause a panic on `insert`.

However, perhaps this should be a private function?

paulhauner commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+        let mut file = File::create(self.filepath.as_path())?;
+
+        // Locking the file to make sure we're atomically writing to it
+        file.lock_exclusive()?;
```
According to the [docs](https://docs.rs/fs2/0.4.3/fs2/trait.FileExt.html#tymethod.lock_exclusive), this will block the current thread until it can obtain a lock on the file.

If there is contention for the file then one of those process will just block here. This seems quite reasonable in the best-case; each process won't have to block for long. However, if one of the processes crashes between unlocking and locking the file (e.g., crash, power failure), then future threads may just block here quietly forever. It's worth noting that this is a OS-level lock and may be triggered from outside of our process or outlive it.

For our application, I think using [`try_lock_exclusive`](https://docs.rs/fs2/0.4.3/fs2/trait.FileExt.html#tymethod.try_lock_exclusive) and immediately returning an error is ideal. It means we'll push the "retry" logic back up into the validator client, but I think that's fine. There we can use timer `futures` to keep polling a locked file, potentially logging warnings if we detect contention.

Also, I think there's a race condition here because we lock the file when `verifying`, instead of locking the file for the entire lifetime of `HistoryInfo` (i.e., taking the lock in the constructor and unlocking it on `Drop`). Consider two processes both accessing the same file:

1. Process `A` opens an existing `HistoryInfo`, loads `n` previous votes into memory.
1. Process `B` does the same.
1. Process `A` calls `update_if_valid(v_1)`.
1. Process `B` calls `update_if_valid(v_2)`.
1. `v_1` is not in the history.

paulhauner commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+        // Locking the file to make sure we're atomically writing to it
+        file.lock_exclusive()?;
+
+        // Setting permissions to be 600 (rw-------).
```
Nice.

If we split into two constructor methods, perhaps we only need to _set_ permissions on `create`.

I was considering perhaps checking for 600 permissions when opening the file and returning an error if they're too loose, but I think this is a bit overboard. I think the file is private enough that we _set_ tight permissions, but not that private that we _enforce_ permissions.

paulhauner commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+impl SafeFromSlashing<SignedAttestation> for HistoryInfo<SignedAttestation> {
+    type U = AttestationData;
+
+    fn verify_and_get_index(
```
Perhaps we shouldn't include this in the public interface?

It's a user-error risk having a function that checks votes but does not remember them. I think we should avoid it, if we can.

paulhauner commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+        self.data.insert(index, data);
+
+        // Creating file if it doesn't exist, else opening it.
+        let mut file = File::create(self.filepath.as_path())?;
```
As mentioned in a later comment, I think we should be explicit about `create`.

paulhauner commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,378 @@
+
+        // Setting permissions to be 600 (rw-------).
+        let mut perm = file.metadata()?.permissions();
+        perm.set_mode(0o600);
```
Cool, I didn't know Rust had octal!

pscott commented:
I've review the comments and implemented the changes.
I've also changed the code in service.rs to reflect the changes, but have NOT included the logic that says "if we failed because of the file lock, try again". To do that, we would probably need two things:
1. Have the safe_to_produce method return an enum instead of a bool (to know if it failed because of the lock or for another reason
2. Determine what exactly we wish to do. Should we try x times? Or wait 10secondes and retry? Or something else?

paulhauner commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,402 @@
+
+        // Writing new history to file
+        self.file.seek(SeekFrom::Start(0))?;
+        self.file.write_all(&self.data.as_ssz_bytes())?;
```
We do have the risk here that if you crash during this operation you may corrupt the history.

There are two ways to deal with this:

1. Roll our own rollback system (e.g., write to a temp file, then copy it over the existing main file).
1. Instead of over-writing your history file, just append to it. This will make the on-disk file unordered, however you could just order it when you put it into memory.
1. Use an existing flat-file, low-throughput database implementation (e.g., SQLite).

SQLite already deals with journals, rollbacks, etc. It would also make writes something less than `O(v)` where `v` is all votes. It will have some space and time overhead, but given the history file is already pretty small, I think that's ok.

I probably should have realized this earlier, but I think it's worth having a look at using [`rusqlite`](https://docs.rs/rusqlite/0.20.0/rusqlite/). If it can be implemented easily (say, ~3 days) and doesn't have any show-stopping problems, this might be the best way to go.

For reference, checkout out [this doc](https://www.sqlite.org/howtocorrupt.html) on "how to corrupt an sqlite database file".

pscott replied:
Done!

pscott commented on validator_client/slashing_protection/src/slashing_protection.rs:
```diff
@@ -0,0 +1,402 @@
+    pub fn open(path: &Path) -> Result<Self, NotSafe> {
+        let mut file = OpenOptions::new().read(true).write(true).open(path)?;
+
+        file.try_lock_exclusive()?;
```
Oh yeah! Regarding `unlock` my understanding is that unlock gets automatically called when the object is dropped. See [this test](https://docs.rs/fs2/0.4.3/src/fs2/lib.rs.html#277) that tests this functionality and [this comment](https://docs.rs/fs2/0.4.3/src/fs2/lib.rs.html#28) that specifies that locks are life-long lasting. However I haven't been able to find an `impl Drop` in the fs2 source code...

paulhauner replied:
Oh you're right. Turns out the lock is dropped on the OS level (at least for [flock](https://linux.die.net/man/2/flock)):

> Furthermore, the lock is released either by an explicit LOCK_UN operation on any of these duplicate descriptors, or when all such descriptors have been closed.

pscott commented:
@paulhauner I just updated this PR with the rusql implementation. HistoryInfo now keeps a connection to the database, and updates the db when needed. The data in the db gets loaded when the HistoryInfo is created. After that, when updating the history, we simply write the new data to the  sql db and update the in-memory Vec. Let me know if this is not valid, or not a good architecture.

Thanks for your feedback! :)

pscott commented:
@paulhauner PR has been updated to use only "one source of truth", i.e the db. Tests are now passing, comments have been updated: awaiting for your review! :D

paulhauner commented:
I just changed the base of this PR to make the diff more manageable :)

paulhauner commented:
As mentioned in chat, here's my branch which solves the `Send` problem using a connection pool. It doesn't compile though because theres a mutability issue.

https://github.com/sigp/lighthouse/tree/paul-slash-protection

paulhauner commented:
Oh, also we generally name a file after the name of the main struct (it's subjective) in that file. Following that, I'd say that the `attester_slashings.rs` file should be named `signed_attestation.rs`. There's a couple more files like this :)

pscott commented on validator_client/src/validator_store.rs:
```diff
@@ -154,15 +219,39 @@ impl<T: SlotClock + 'static, E: EthSpec> ValidatorStore<T, E> {
+        // Retrieving the corresponding ValidatorDir
+        let validator = match validators.get(validator_pubkey) {
+            Some(validator) => validator,
+            None => return None,
```
Maybe we should log something here, stating that the validator_pubkey doesn't appear in amongst the validators?

pscott commented on validator_client/src/validator_store.rs:
```diff
@@ -154,15 +219,39 @@ impl<T: SlotClock + 'static, E: EthSpec> ValidatorStore<T, E> {
+        let is_slashing_free = validator
+            .block_slashing_protection
+            .as_ref()?
+            .try_lock()?
```
What should we do here if it fail? Need to check upstream which behaviour we want?

pscott commented on validator_client/src/validator_store.rs:
```diff
@@ -171,30 +260,521 @@ impl<T: SlotClock + 'static, E: EthSpec> ValidatorStore<T, E> {
+        let is_slashing_free = validator
+            .attestation_slashing_protection
+            .as_ref()?
+            .try_lock()?
```
What should we do here if it fail? Need to check upstream which behaviour we want?

paulhauner commented:
I'm still keen to get this merged, I'm just dealing with some more immediate sync-speed issues at the moment.

Please bear with me! :)

michaelsproul commented:
Closing in favour of #1116, thanks @pscott for laying the groundwork! :tada:
